Audit job to fetch invalid `tabs` and `collapsibles` data

lkbhitesh07 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: Fetches all the invalid data from the tabs and collapsible RTE tags.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

lkbhitesh07 committed:
initial changes for tabs and collapsibles

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
added code for tabs and collapsibles

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
tested the file

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
made changes

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
fixed few things

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
added some functionality

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 commented:
@EricZLou PTAL

lkbhitesh07 commented:
Hey @EricZLou,
Please note that the backend tests will fail because of the sequence of the resulting data. This is happening because when I iterate over RTE tags present inside the `tabs` or `collapsible` I'm using `soup.find_all()` which do not iterate over the tags in sequence and when that happens the error related to the tags are stored in different sequence every time.
Thanks

EricZLou commented:
> 

Noted -- but when the ordering is the same, then the backend test would pass, right?

EricZLou commented on core/jobs/batch_jobs/tabs_collapsables_validation_jobs.py:
```diff
@@ -0,0 +1,582 @@
+        for tag in soup.find_all():
+            if tag.name == 'oppia-noninteractive-image':
+                try:
+                    if tag['alt-with-value'] in (
+                        '&quot;&quot;', '', "", '\'\'', '\"\"'):
```
In Python, `''` (2 single quotes) and `""` (2 double quotes) are the same thing... both empty strings. `'\'\''` and `'\"\"'` are different of course because that's a string that contains 2 single quotes vs a string that contains 2 double quotes

lkbhitesh07 replied:
My bad, I forgot to remove them as I was testing different things. I will do that.
Thanks for noticing it.

lkbhitesh07 replied:
Done

EricZLou commented on core/jobs/batch_jobs/tabs_collapsables_validation_jobs.py:
```diff
@@ -0,0 +1,582 @@
+                    )
+
+            elif tag.name == 'oppia-noninteractive-video':
+                try:
+                    tag['start-with-value']
+
+                except Exception:
+                    states_with_errored_values.append(
+                        'start value invalid'
+                    )
```
Lol I don't want to repeat the same comments as previous PRs but the right thing to do is:
```
if 'start-with-value' not in tag:
    ... append error ...
```

lkbhitesh07 replied:
I tried doing that previously, we won't be able to do that because here the `tag` is not considered as a string it is a value that is of type from `bs4`. It will always go the `else` part.
Thanks

EricZLou replied:
Ah I see thanks for clarifying. Bs4 might have a class method that does this operation?

lkbhitesh07 replied:
Yeah

EricZLou replied:
Have you found if there's a method for this?

lkbhitesh07 replied:
Yes, the type of the tag is `bs4.element.Tag`

EricZLou replied:
No, I mean Bs4 may have a cleaner method to call. Like tag.contains('start-with-value'). So then instead of try except you can do an if statement, which is better Python.

lkbhitesh07 replied:
I see, my bad. I have found one method which is `tag.has_attr()`. Will it be okay if I use this thing in the conversion function instead, otherwise there will be lots of changes?
Thanks

EricZLou reviewed (changes requested):
Aren't there more RTE checks than this? Does this check everything on the RTE tab in the Android validation spreadsheet? https://docs.google.com/spreadsheets/d/1kNLF-UkS_Qk0igNCGwLjJXM0ByHalqli41GPvWF43PE/edit#gid=259309225

lkbhitesh07 commented:
> > 
> 
> Noted -- but when the ordering is the same, then the backend test would pass, right?

Though I will try something that I have on my mind which can help us to pass the test. I will update you today.

lkbhitesh07 commented:
> Aren't there more RTE checks than this? Does this check everything on the RTE tab in the Android validation spreadsheet? https://docs.google.com/spreadsheets/d/1kNLF-UkS_Qk0igNCGwLjJXM0ByHalqli41GPvWF43PE/edit#gid=259309225

I see, there are 2-3 checks that are remaining in this part. I will update you by today.
Thanks @EricZLou

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
added tests and RTE restricted check

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 commented:
I have added the check for the curated exps to only allow image, skillreview and math tag. I have also fixed the backend test cases to pass @EricZLou PTAL.
Thanks

EricZLou commented on core/jobs/batch_jobs/tabs_collapsables_validation_jobs.py:
```diff
@@ -147,25 +158,40 @@ def _filter_invalid_rte_values(html, states_with_errored_values):
 
                 try:
-                    if tag['autoplay-with-value'] not in ('true', 'false'):
+                    if tag['autoplay-with-value'] not in (
+                        'true', 'false', '\"true\"', '\"false\"', True, False):
```
I'm guessing you did this because you haven't looked into what the possible values can be? I get what you're doing and I guess it's fine, but in the future, look into it first and then write the code.

lkbhitesh07 replied:
I have looked at the values and the values can only be of the form `"true" or "false"` but Sean mentioned previously that if there is any possibility from which we can depict the original intention of the user we should do so.

U8NWXD commented:
@lkbhitesh07 I see that you have failing [lint checks](https://github.com/oppia/oppia/runs/8287415621?check_suite_focus=true), which should have run on your local machine when you pushed your changes. Did these checks run locally? (I want to make sure our pre-push hooks are working!)

lkbhitesh07 commented:
> @lkbhitesh07 I see that you have failing [lint checks](https://github.com/oppia/oppia/runs/8287415621?check_suite_focus=true), which should have run on your local machine when you pushed your changes. Did these checks run locally? (I want to make sure our pre-push hooks are working!)

Hey @U8NWXD, Yes but I used `--no-verify` to push the changes for this part. The local checks works fine.
Thanks

lkbhitesh07 committed:
added one more test

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
updated some code related to develop

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

kevintab95 commented:
Hi @lkbhitesh07, what is the next step on this PR?

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
updated for desired results

lkbhitesh07 committed:
fixed backend import test

lkbhitesh07 commented:
This was a one-off job that required gathering invalid data. I have received the results hence closing this.
Thanks

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into tabs_collapsable

lkbhitesh07 committed:
made changes to log the errors

lkbhitesh07 committed:
nit
